colosseum/declaration.py

```diff
@@ -1,33 +1,108 @@
+import collections
+
 from . import engine as css_engine
 from .constants import (  # noqa
     ALIGN_CONTENT_CHOICES, ALIGN_ITEMS_CHOICES, ALIGN_SELF_CHOICES, AUTO,
     BACKGROUND_COLOR_CHOICES, BORDER_COLOR_CHOICES, BORDER_STYLE_CHOICES,
     BORDER_WIDTH_CHOICES, BOX_OFFSET_CHOICES, CLEAR_CHOICES, COLOR_CHOICES,
     DIRECTION_CHOICES, DISPLAY_CHOICES, FLEX_BASIS_CHOICES,
     FLEX_DIRECTION_CHOICES, FLEX_GROW_CHOICES, FLEX_SHRINK_CHOICES, FLEX_START,
-    FLEX_WRAP_CHOICES, FLOAT_CHOICES, GRID_AUTO_CHOICES,
-    GRID_AUTO_FLOW_CHOICES, GRID_GAP_CHOICES, GRID_PLACEMENT_CHOICES,
-    GRID_TEMPLATE_AREA_CHOICES, GRID_TEMPLATE_CHOICES, INLINE,
-    JUSTIFY_CONTENT_CHOICES, LTR, MARGIN_CHOICES, MAX_SIZE_CHOICES,
-    MIN_SIZE_CHOICES, NORMAL, NOWRAP, ORDER_CHOICES, PADDING_CHOICES,
-    POSITION_CHOICES, ROW, SIZE_CHOICES, STATIC, STRETCH,
-    TRANSPARENT, UNICODE_BIDI_CHOICES, VISIBILITY_CHOICES, VISIBLE,
-    Z_INDEX_CHOICES, default,
+    FLEX_WRAP_CHOICES, FLOAT_CHOICES, FONT_FAMILY_CHOICES, FONT_SIZE_CHOICES,
+    FONT_STYLE_CHOICES, FONT_VARIANT_CHOICES, FONT_WEIGHT_CHOICES,
+    GRID_AUTO_CHOICES, GRID_AUTO_FLOW_CHOICES, GRID_GAP_CHOICES,
+    GRID_PLACEMENT_CHOICES, GRID_TEMPLATE_AREA_CHOICES, GRID_TEMPLATE_CHOICES,
+    INITIAL, INLINE, JUSTIFY_CONTENT_CHOICES, LINE_HEIGHT_CHOICES, LTR,
+    MARGIN_CHOICES, MAX_SIZE_CHOICES, MEDIUM, MIN_SIZE_CHOICES, NORMAL, NOWRAP,
+    ORDER_CHOICES, PADDING_CHOICES, POSITION_CHOICES, ROW, SIZE_CHOICES,
+    STATIC, STRETCH, TRANSPARENT, UNICODE_BIDI_CHOICES, VISIBILITY_CHOICES,
+    VISIBLE, Z_INDEX_CHOICES, default, INITIAL_FONT_VALUES, EMPTY,
 )
+from .exceptions import ValidationError
+from .parser import parse_font
+from .wrappers import FontFamily, FontShorthand
 
 _CSS_PROPERTIES = set()
 
 
-def unvalidated_property(name, choices, initial):
-    "Define a simple CSS property attribute."
-    initial = choices.validate(initial)
+def validated_shorthand_property(name, initial, parser, storage_class):
+    """Define the shorthand CSS font property."""
+    initial = storage_class(**initial)
+
+    def getter(self):
+        shorthand = initial.copy()
+        for property_name in shorthand:
+            shorthand[property_name] = getattr(self, property_name)
+        return shorthand
+
+    def setter(self, value):
+        try:
+            shorthand = storage_class(**parser(value))
```

Comment:
@freakboy3742 what should be the behavior when setting a font shorthand?

If the user only provides some parts, let's say"

style.font = "12px/3 serif"

Since this shorthand resets the missing properties to their initial values, then should we explicitly write the properties or delete them (which would be akin to setting them to their initial values when the user requests them)

So:

Set properties with missing values
```
style.font = "12px/3 serif"
print(style)
font: ..
font_size: ...
font_family: ...
font_style: ...
font_variant: ...
font_weight: ...
```
vs.

Delete properties with missing values
```
style.font = "12px/3 serif"
print(style)
font: ..
font_size: ...
font_family: ...
```

Reply:
I'd say the latter. "normal normal normal normal 12px/3 serif" doesn't communicate anything that "12px/3 serif" doesn't communicate; and I'd say the same is also true for the inverse.  If I explicitly ask for "what is the font variant", then the answer is "normal"; but if I just ask for the font, the explicitness of including "normal" for font variant doesn't really aid any understanding.

Reply:
👍